src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs

```diff
@@ -34,14 +34,26 @@ public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeP
                 return;
             }
 
+            DateTimeOffset stamp;
+            if (logEntry.State is IBufferedLogRecord bufferedLogRecord)
```

Comment:
I have a couple questions...

* Will this test cause all `struct` `TState`s to be boxed?

* If some `ILoggerProvider` doesn't support `IBufferedLogRecord` what happens? Timestamps end up incorrect and correlation breaks?

Could we somehow make it an opt-in thing? Thinking something like...

```csharp
public interface ISupportBufferedLogging
{
        void Log(in BufferedLogEntry log);
}

public readonly ref struct BufferedLogEntry
{
   public string CategoryName { get; }
   public DateTimeOffset Timestamp { get; }
   public LogLevel LogLevel { get; }
   public EventId EventId { get; }
   public Exception? Exception { get; }
   public ActivityContext ActivityContext { get; }
   public string? FormattedMessage { get; }
   public string? MessageTemplate { get; }
   public ReadOnlySpan<KeyValuePair<string, object?>> Attributes { get; }
}
```

Reply:
@stephentoub Does the pattern here lead to unconditional boxing? Is there a workaround to prevent that? I seem to recall there's a magic sequence that the JIT recognize which will make it not box...

Reply:
@CodeBlanch So you're proposing making so the type test happens in the infrastructure? So before calling ILogger.Log on a registered logger, the infra would do a type test there and if its available it would call the new call you show? @noahfalk This actually seems like a pretty clean approach. WDYT?

And yes, if the user starts to use buffering and sends data to a non-upgraded log provider, the time stamp that the log provider will use will be when the log records were flushed, rather than when they were created.

Reply:
@geeknoid 

> So you're proposing making so the type test happens in the infrastructure? So before calling ILogger.Log on a registered logger, the infra would do a type test there and if its available it would call the new call you show?

Ya that's what I was thinking. The `ILoggerFactory` would be able to reason out if any/all the `ILoggerProvider`s registered support buffering and do the right thing based on what opts-in using the interface, whatever "the right thing" is 🤣 Could be for non-supporting `ILoggerProvider` they just get the logs as they did before. Could be a throw when things are mixed. Not sure what that implementation needs to be exactly.

There's also a couple other things I tossed in there. `BufferedLogEntry` takes care of handling/parsing `TState` into something more well-defined. Today every `ILoggerProvider` has to reason out what to do with `TState` this could be a good opportunity to fix that. It is also `ref struct` so that it can expose `ReadOnlySpan<KeyValuePair<string, object?>>` for the `attributes`. The idea being `ILoggerProvider` implementations supporting `ISupportBufferedLogging` will need to a) fully consume that `BufferedLogEntry` when `Log` is called or b) copy it over to some kind of pooled storage for delayed sending. The `ref struct` prevents people from trying to box it or hold onto it which allows the buffering implementation to pool these things safely without worry.

Reply:
@CodeBlanch Given your description, this would be desirable even for non-buffering cases. Basically, this is a new contract between the infra and the logging providers which has higher-level semantics, ensuring more consistency between log providers, while also delivering the needed changes to properly support buffering.

Reply:
This feels like you are proposing ILoggerV2? I understand the appeal as I spent a bunch of time jousting at that design problem during .NET 8, but I also know from that experience the complexity can creep up fast. It is not a design can of worms I think we should open at this stage in a release. To me ILoggerV2 feels like something we get one shot at, then we spend 5+ years in a transitory period where everything is more complicated because both V1 and V2 are supported, and only if/when we finally start removing support for ILoggerV1 does anything actually get simpler.

If folks are looking for some kind of error handling when buffering is in use I'd suggest first lets define what behavior is desired before we try to design an API. For example if the desired behavior is 'pass it through anyways' then we don't need any changes for that. If the desired behavior is 'drop the messages', something like this would probably do the job.
```
ILoggerProviderSupportsBuffering { bool SupportsBuffering { get; } }
```

Reply:
@noahfalk 

> This feels like you are proposing ILoggerV2?

I wouldn't say that. No changes are being suggested to the APIs used to log anything. Users will continue to log as they always have. For _providers_, if they want to support buffering, they will need to make changes. But this is a tiny set of users/components.

Aside from the niceness of the API, there are other reasons why I think we consider doing this 🤔

When we talk about **THE** logging API (there is only one at the end of the day):

```csharp
    public interface ILogger
    {
       void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter);
    }
```

There are no restrictions placed upon `TState` whatsoever.

Users may do this if they want:

```csharp
    public async Task UpdateUser(User user, UpdateUserRequest request)
    {
        logger.Log<User>(LogLevel.Debug, eventId: default, state: user, exception: null, formatter: user.LogUserBeforeChange);

        await UpdateUser(user, request);

        logger.Log<User>(LogLevel.Debug, eventId: default, state: user, exception: null, formatter: user.LogUserAfterChange);

        logger.LogInformation("User changed {userId}", user.UserId);
    }
```

What is being demoed there is logging something which is mutable (`User`). This is perfectly valid.

When it comes to buffering, we shouldn't just hold onto `TState` and then hand it out later when we decide it should be logged. It could have mutated by then. What the buffering should do (IMO) is before `Log` returns, transfer the data onto something immutable.

If we don't do that, users may see data in the log which was changed after the log was expected to be written. Or worse, we may try to write something later which has been disposed and throws. This is a real issue we ran into in OTel (which already has pooling/buffering/batching): https://github.com/open-telemetry/opentelemetry-dotnet/issues/2905

Reply:
Yes, buffering involves serializing the state in memory. The original TState cannot be kept and reused at some random time in the future, it must be fully consumed by the time ILogger.Log returns.